pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -176,6 +176,18 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 	// Read the initial condition
 	before := pr.Status.GetCondition(apis.ConditionSucceeded)
 
+	// Check if we are failing to mark this as timed out for a while. If we are, mark immediately and finish the
+	// reconcile
+	if pr.HasTimedOutForALongTime(ctx, c.Clock) && !isPipelineRunTimeoutConditionSet(pr) {
+		setPipelineRunTimeoutCondition(ctx, pr)
```

Comment:
Why not call `timeoutPipelineRun` instead to ensure the child objects have been attempted at least?
That won't cause an increase in the size of `PipelineRun` and will behave better in cases where the long timeout is caused by a different situation like for instance the controller being unresponsive or down for some time.

The only extra bit of data that `timeoutPipelineRun` would write is the completion time, which is unlikely to cause the etcd failure I think. If you would prefer to not write a completion time, we should make it obvious in the comment what is going on.

Reply:
I didn't thought about it, to be honest. Fair enough.